app/views/tutorial/a3.html

```diff
@@ -0,0 +1,243 @@
+                                    <td>&lt;div style="width:
+                                        <span style="color:red;">UNTRUSTED DATA</span>;"&gt;Selection&lt;/div&gt;</td>
+                                    <td>Except for alphanumeric characters, escape all characters with ASCII values less
+                                        than 256 with the \HH (HH= Hex Value) escaping format.
+                                    </td>
+                                </tr>
+                            </tbody>
+                        </table>
+                    </li>
+                    <li>
+                        <p><b>HTTPOnly cookie flag:</b> Preventing all XSS flaws in an application is hard. To help mitigate
+                            the impact of an XSS flaw on your site, set the HTTPOnly flag on session cookie and any custom
+                            cookies that are not required to be accessed by JavaScript.
+                        </p>
+                    </li>
+                    <li>
+                        <p><b>Implement Content Security Policy (CSP):</b> CSP is a browser side mechanism which allows creating
+                            whitelists for client side resources used by the web application, e.g. JavaScript, CSS, images,
+                            etc. CSP via special HTTP header instructs the browser to only execute or render resources from
+                            those sources. For example, the CSP header below allows content only from example site's own
+                            domain (mydomain.com) and all its sub domains.
+                            <pre>Content-Security-Policy: default-src 'self' *.mydomain.com</pre>
+
+                        </p>
+                    </li>
+                    <li> <b>Apply encoding on both client and server side: </b> It is essential to apply encoding on both
+                        client and server side to mitigate DOM based XSS attack, in which untrusted data never leaves the
+                        browser.
+                </ol>
+                <p>Source: XSS Prevention Cheat Sheet[1]
+                </p>
+            </div>
+        </div>
+        <div id="source-code-example" class="panel panel-info">
+            <div class="panel-heading">
+                <h3 class="panel-title">Source Code Example</h3>
+            </div>
+            <div class="panel-body">
+                <p>
+                    The demo web application is vulnerable to stored XSS attack on profiles form. On form submit, the first and last name field
+                    values are submitted to the server, and without any validation get saved in database. The values are
+                    then sent back to the browser without proper escaping to be shown at the top right menu.
+                </p>
+                <iframe width="560" height="315" src="//www.youtube.com/embed/KvZ5jdg083M?rel=0" frameborder="0" allowfullscreen></iframe>
```

Comment:
#### Automated PR Comment From Polaris SAST
<!-- da79dbd4aa99fe12946d42da87f54b6a -->
<!-- Comment managed by Git Commenter, do not modify! -->
### Polaris SAST Issue - Inclusion of Functionality from Untrusted Control Sphere
_Medium CWE-829_
The `sandbox` attribute of an `iframe` is not set in HTML. This iframe loads content from a third-party site. If the JavaScript in that iframe is malicious or was compromised, an attacker can break out of the iframe, access the DOM of the main page, and mount a clickjacking or a phishing attack.



### How to fix
Potential mitigations include the following:
  1. Use a vetted library or framework that does not allow this weakness to occur or provides constructs that make this weakness easier to avoid.
  2. When the set of acceptable objects, such as filenames or URLs, is limited or known, create a mapping from a set of fixed input values (such as numeric IDs) to the actual filenames or URLs, and reject all other inputs.

For example, ID 1 could map to "inbox.txt" and ID 2 could map to "profile.txt". Features such as the ESAPI AccessReferenceMap [REF-45] provide this capability.

  3. For any security checks that are performed on the client side, ensure that these checks are duplicated on the server side, in order to avoid CWE-602. Attackers can bypass the client-side checks by modifying values after the checks have been performed, or by changing the client to remove the client-side checks entirely. Then, these modified values would be submitted to the server.
  4. Run the code in a "jail" or similar sandbox environment that enforces strict boundaries between the process and the operating system. This may effectively restrict which files can be accessed in a particular directory or which commands can be executed by the software.

OS-level examples include the Unix chroot jail, AppArmor, and SELinux. In general, managed code may provide some protection. For example, java.io.FilePermission in the Java SecurityManager allows the software to specify restrictions on file operations.

This may not be a feasible solution, and it only limits the impact to the operating system; the rest of the application may still be subject to compromise.

Be careful to avoid CWE-243 and other weaknesses related to jails.

  5. Run your code using the lowest privileges that are required to accomplish the necessary tasks [REF-76]. If possible, create isolated accounts with limited privileges that are only used for a single task. That way, a successful attack will not immediately give the attacker access to the rest of the software or its environment. For example, database applications rarely need to run as the database administrator, especially in day-to-day operations.
  6. Assume all input is malicious. Use an "accept known good" input validation strategy, i.e., use a list of acceptable inputs that strictly conform to specifications. Reject any input that does not strictly conform to specifications, or transform it into something that does.

When performing input validation, consider all potentially relevant properties, including length, type of input, the full range of acceptable values, missing or extra inputs, syntax, consistency across related fields, and conformance to business rules. As an example of business rule logic, "boat" may be syntactically valid because it only contains alphanumeric characters, but it is not valid if the input is only expected to contain colors such as "red" or "blue."

Do not rely exclusively on looking for malicious or malformed inputs. This is likely to miss at least one undesirable input, especially if the code's environment changes. This can give attackers enough room to bypass the intended validation. However, denylists can be useful for detecting potential attacks or determining which inputs are so malformed that they should be rejected outright.

When validating filenames, use stringent allowlists that limit the character set to be used. If feasible, only allow a single "." character in the filename to avoid weaknesses such as CWE-23, and exclude directory separators such as "/" to avoid CWE-36. Use a list of allowable file extensions, which will help to avoid CWE-434.

Do not rely exclusively on a filtering mechanism that removes potentially dangerous characters. This is equivalent to a denylist, which may be incomplete (CWE-184). For example, filtering "/" is insufficient protection if the filesystem also supports the use of "\" as a directory separator. Another possible error could occur when the filtering is applied in a way that still produces dangerous data (CWE-182). For example, if "../" sequences are removed from the ".../...//" string in a sequential fashion, two instances of "../" would be removed from the original string, but the remaining characters would still form the "../" string.

  7. Store library, include, and utility files outside of the web document root, if possible. Otherwise, store them in a separate directory and use the web server's access control capabilities to prevent attackers from directly requesting them. One common practice is to define a fixed constant in each calling program, then check for the existence of the constant in the library/include file; if the constant does not exist, then the file was directly requested, and it can exit immediately.

This significantly reduces the chance of an attacker being able to bypass any protection mechanisms that are in the base program but not in the include files. It will also reduce the attack surface.

  8. Understand all the potential areas where untrusted inputs can enter your software: parameters or arguments, cookies, anything read from the network, environment variables, reverse DNS lookups, query results, request headers, URL components, e-mail, files, filenames, databases, and any external systems that provide data to the application. Remember that such inputs may be obtained indirectly through API calls.

Many file inclusion problems occur because the programmer assumed that certain inputs could not be modified, especially for cookies and URL components.

  9. Use an application firewall that can detect attacks against this weakness. It can be beneficial in cases in which the code cannot be fixed (because it is controlled by a third party), as an emergency prevention measure while more comprehensive software assurance measures are applied, or to provide defense in depth.